src/Logger/Logger.php

```diff
@@ -0,0 +1,364 @@
+     * @return Logger
+     */
+    public function setLogLevel(int $level): Logger
+    {
+        $levels         = $this->getLevels();
+        $this->logLevel = isset($levels[$level]) ? $level : self::CUSTOM;
+
+        return $this;
+    }
+
+    /**
+     * Adds a message to each handler for processing
+     *
+     * @param int    $level
+     * @param string $message
+     * @param array  $context
+     *
+     * @return bool
+     * @throws Exception
+     * @throws LoggerException
+     */
+    protected function addMessage(
+        int $level,
+        string $message,
+        array $context = []
+    ): bool {
+        if ($this->logLevel >= $level) {
+            if (count($this->adapters) === 0) {
+                throw new LoggerException('No adapters specified');
+            }
+
+            $levels    = $this->getLevels();
+            $levelName = $levels[$level] ?? self::CUSTOM;
+
+            $item = new Item(
+                $message,
+                $levelName,
+                $level,
+                new DateTimeImmutable('now', $this->timezone),
+                $context
+            );
+
+            /**
+             * Log only if the key does not exist in the excluded ones
+             */
+            $collection = array_diff_key($this->adapters, $this->excluded);
+            foreach ($collection as $adapter) {
+                $method = 'process';
+                if (true === $adapter->inTransaction()) {
+                    $method = 'add';
+                }
+
+                $adapter->$method($item);
+            }
+
+            /**
+             * Clear the excluded array since we made the call now
+             */
+            $this->excluded = [];
+        }
+
+        return true;
+    }
+
+    /**
+     * Returns an array of log levels with integer to string conversion
+     */
+    protected function getLevels(): array
+    {
+        return [
+            self::ALERT     => 'alert',
+            self::CRITICAL  => 'critical',
+            self::DEBUG     => 'debug',
+            self::EMERGENCY => 'emergency',
+            self::ERROR     => 'error',
+            self::INFO      => 'info',
+            self::NOTICE    => 'notice',
+            self::WARNING   => 'warning',
+            self::CUSTOM    => 'custom',
+        ];
+    }
+
+    /**
+     * Converts the level from string/word to an integer
+     *
+     * @param mixed $level
+     *
+     * @return int
+     */
+    private function getLevelNumber($level): int
+    {
+        if (is_string($level)) {
+            $levelName = mb_strtolower($level);
+            $levels    = array_flip($this->getLevels());
+
+            if (isset($levels[$levelName])) {
+                return $levels[$levelName];
+            }
+        } elseif (is_numeric($level)) {
+            $levels = $this->getLevels();
+
+            if (isset($levels[$level])) {
+                return (int) $level;
+            }
+        }
+
+        return self::CUSTOM;
```

Comment:
The specification says: 
> Calling this method with a level not defined by this specification MUST throw a Psr\Log\InvalidArgumentException if the implementation does not know about the level. Users SHOULD NOT use a custom level without knowing for sure the current implementation supports it.

This currently defaults to `CUSTOM` if the level is not known by the adapter.